MapboxNavigation/NavigationMapView.swift

```diff
@@ -469,62 +468,101 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
-        let mainRouteCasingSource = addMainRouteCasingSource(style, shape: mainRouteCasingShape)
-        
-        let mainRouteLayer = addMainRouteLayer(style, source: allRoutesSource, lineGradient: routeLineGradient(mainRoute, fractionTraveled: 0.0))
-        let mainRouteCasingLayer = addMainRouteCasingLayer(style, source: mainRouteCasingSource, lineGradient: routeCasingGradient(0.0), below: mainRouteLayer)
-        let alternativeRoutesLayer = addAlternativeRoutesLayer(style, source: allRoutesSource, below: mainRouteCasingLayer)
-        addAlternativeRoutesCasingLayer(style, source: allRoutesSource, below: alternativeRoutesLayer)
+        var parentLayer: MGLStyleLayer? = nil
+        for (index, route) in routes.enumerated() {
+            guard let routeSourceIdentifier = identifier(route, identifierType: .source),
+                  let routeCasingSourceIdentifier = identifier(route, identifierType: .source, isMainRouteCasingSource: true),
+                  let routeIdentifier = identifier(route, identifierType: .route),
+                  let routeCasingIdentifier = identifier(route, identifierType: .routeCasing) else { continue }
+            
+            // In case of main route there is the ability to provide custom `MGLShape` for either route or route casing by implemeting
+            // `NavigationMapViewDelegate.navigationMapView(_:shapeFor:)` or `NavigationMapViewDelegate.navigationMapView(_:simplifiedShapeFor:)`.
+            if index == 0 {
+                let routeShape = navigationMapViewDelegate?.navigationMapView(self, shapeFor: [route]) ??
+                    shape(for: route, legIndex: legIndex, isAlternateRoute: false)
+                
+                let routeSource = addRouteSource(style, identifier: routeSourceIdentifier, shape: routeShape)
+                
+                let mainRouteLayer = addMainRouteLayer(style,
+                                                       source: routeSource,
+                                                       identifier: routeIdentifier,
+                                                       lineGradient: routeLineGradient(route, fractionTraveled: 0.0))
+                
+                let mainRouteCasingShape = navigationMapViewDelegate?.navigationMapView(self, simplifiedShapeFor: route) ??
+                    shape(forCasingOf: route, legIndex: legIndex)
+                
+                let routeCasingSource = addRouteSource(style, identifier: routeCasingSourceIdentifier, shape: mainRouteCasingShape)
+                
+                parentLayer = addMainRouteCasingLayer(style,
+                                                      source: routeCasingSource,
+                                                      identifier: routeCasingIdentifier,
+                                                      lineGradient: routeCasingGradient(0.0),
+                                                      below: mainRouteLayer)
+                
+                continue
```

Comment:
I think this is the most important change. As documentation was slightly changed in https://github.com/mapbox/mapbox-navigation-ios/pull/2623 to address that `NavigationMapViewDelegate.navigationMapView(_:shapeFor:)` and `NavigationMapViewDelegate.navigationMapView(_:simplifiedShapeFor:)` are meant to be used only for main route in comment:
```
Resulting `MGLShape` will then be styled using `NavigationMapView.navigationMapView(_: mainRouteStyleLayerWithIdentifier: source:)` provided style or a default congestion style if above delegate method was not implemented. In latter case, consider modifing your custom `MGLShape` `attributes` to have 'isAlternateRoute' key set to 'false'. Otherwise style predicate condition will filter out the shape.
```
I think it has become more misleading, as `NavigationMapViewDelegate.navigationMapView(_:shapeFor:)` allows to change shape not only for main route, but for all routes.

After adding the ability to provide route style for every route in this PR `NavigationMapViewDelegate.navigationMapView(_:shapeFor:)` and `NavigationMapViewDelegate.navigationMapView(_:simplifiedShapeFor:)` will be called only for main route, this means there will be no ability to provide custom shape for alternative routes and their casings.

Because of this I think `NavigationMapViewDelegate.navigationMapView(_:shapeFor:)` can be updated to be similar to `NavigationMapViewDelegate.navigationMapView(_:simplifiedShapeFor:)` and provide `Route` instead of `[Route]`, but it'd break public API.

@1ec5, @Udumft please let me know if that sounds reasonable and any changes are needed.

Example app was also updated to showcase the ability to provide custom layers and shapes in https://github.com/mapbox/navigation-ios-examples/pull/83.

Reply:
> Because of this I think `NavigationMapViewDelegate.navigationMapView(_:shapeFor:)` can be updated to be similar to `NavigationMapViewDelegate.navigationMapView(_:simplifiedShapeFor:)` and provide `Route` instead of `[Route]`, but it'd break public API.

We can add a new delegate method and deprecate the old one, as long as the old one continues to function reasonably appropriately.

Reply:
To play it safe I'm inclined to leave current implementation without adding new/deprecating old APIs: end user will be able to change shape only for main route instead of all routes (main + alternative). This will match current docs as well.